hw 1 load questions from csv, output full question details into conso…

kondury opened the pull request:
…le out

kondury committed:
hw 1 load questions from csv, output full question details into console out

stvort commented on Quiz-v01/src/main/java/com/github/kondury/quiz/io/ConsoleOutputService.java:
```diff
@@ -0,0 +1,15 @@
+
+import java.io.PrintStream;
+
+public class ConsoleOutputService implements OutputService {
```
Лучше, чтобы данный сервис ничего не знал про предметную область (вопросы), а просто умел читать и выводить строки

kondury replied:
сделано

stvort commented on Quiz-v01/src/main/java/com/github/kondury/quiz/io/ConsoleOutputService.java:
```diff
@@ -0,0 +1,15 @@
+
+public class ConsoleOutputService implements OutputService {
+
+    private final PrintStream out = System.out;
```
Причем `PrintStream` лучше передать извне

kondury replied:
сделано

stvort commented on Quiz-v01/src/main/java/com/github/kondury/quiz/io/OutputService.java:
```diff
@@ -0,0 +1,15 @@
+public interface OutputService {
+    void output(final Question question, final QuestionFormatter formatter);
+
+    default void output(final Question question, final QuestionFormatter formatter, final String prefix, final String suffix) {
+        output(question, it -> prefix + formatter.format(it) + suffix);
+    }
+
+    default void output(final Question question) {
+        output(question, Record::toString);
+    }
```
То что не используется, лучше в репозитории не держать

kondury replied:
удалено

stvort commented on Quiz-v01/src/main/java/com/github/kondury/quiz/dao/CsvQuizDao.java:
```diff
@@ -0,0 +1,55 @@
+
+    private static final String CSV_DAO_FILENAME = "questions.csv";
+
+    private final List<Question> questions;
```
Если можно обойтись без состояния, то лучше обойтись. Это поле может быть локальной переменной аргументом методов и возвращаемым типом. Если прямое чтение из файла, когда-нибудь станет проблемой производительности (что вряд ли), мы будем это решать уровнем выше, в сервисе, с помощью предназначенных для этого инструментов (кэшей)

kondury replied:
сделано

stvort commented on Quiz-v01/src/main/java/com/github/kondury/quiz/dao/CsvQuizDao.java:
```diff
@@ -0,0 +1,55 @@
+        return Collections.unmodifiableList(questions);
+    }
+
+    private static class CsvQuestionsLoader {
```
Этот класс не нужен. Либо его надо выносить наружу, либо его методы перенести в родительский

kondury replied:
перенес методы в родительский, в hw 2 перенесу в utils

stvort commented on Quiz-v01/src/main/java/com/github/kondury/quiz/dao/CsvQuizDao.java:
```diff
@@ -0,0 +1,55 @@
+
+    private static class CsvQuestionsLoader {
+
+        public static List<Question> load(String csvDaoFilename) throws IOException {
```
Ну и статические методы лучше без особой нужды не использовать. Чем меньше область действия чего угодно, тем лучше

stvort commented on Quiz-v01/src/main/java/com/github/kondury/quiz/dao/CsvQuizDao.java:
```diff
@@ -0,0 +1,55 @@
+    private final List<Question> questions;
+
+    public CsvQuizDao() throws IOException {
+        questions = CsvQuestionsLoader.load(CSV_DAO_FILENAME);
```
Логику в конструкторе лучше не писать

stvort commented on Quiz-v01/src/main/java/com/github/kondury/quiz/domain/Question.java:
```diff
@@ -0,0 +1,45 @@
+
+public record Question(String content, Answer answer) {
+
+    public static class Factory {
```
Это совершенно точно не часть вопроса. Вопрос это доменная сущность, которая находится на самом нижнем архитектурном слое. Единственная задача доменного объекта, это объектное представление данных.  В теории такая сущность может быть использована в куче разных приложений. В каждом из них она будет получаться из разных источников. Если все сюда добавлять, то сущность будет постоянно расти

stvort commented on Quiz-v01/src/main/java/com/github/kondury/quiz/domain/Answer.java:
```diff
@@ -0,0 +1,5 @@
+package com.github.kondury.quiz.domain;
+
+public interface Answer {
```
Я бы обошелся без интерфейсов в доменных объектах. Ну и ответы не проверяют ответы. Это не их задача, а задача сервиса по проверке ответов на вопрос

kondury replied:
переделал на предложенную ниже схему

stvort commented on Quiz-v01/src/main/java/com/github/kondury/quiz/domain/ChoiceAnswer.java:
```diff
@@ -0,0 +1,15 @@
+import java.util.Map;
+
+public record ChoiceAnswer(Map<String, Boolean> choices) implements Answer {
+    public ChoiceAnswer(Map<String, Boolean> choices) {
+        this.choices = Collections.unmodifiableMap(choices);
+    }
```
Не очень понятно, как варианты могут быть частью ответа. С т.з. предметной области у нас есть вопрос с вариантами ответов, один или несколько из которых правильные или без вариантов, с одним правильным ответом. Т.е. есть вопрос, вариант ответа (правильный/не правильный) и видимо тип вопроса. Получается что-то типа:
```
class Question {
  private String text;
  private QuestionType type; // enum
  private List<Answer> answers;
}

class Answer {
  private String text;
  private boolean correct;
}
```

kondury replied:
переделал на предложенный вариант, в hw 2 планирую приделать стратегии для печати, проверки

stvort commented on Quiz-v01/src/main/java/com/github/kondury/quiz/dao/CsvQuizDao.java:
```diff
@@ -0,0 +1,55 @@
+
+public class CsvQuizDao implements QuizDao {
+
+    private static final String CSV_DAO_FILENAME = "questions.csv";
```
>4. Имя ресурса с вопросами (CSV-файла) необходимо захардкодить строчкой в XML-файле с контекстом.

kondury replied:
сделано

stvort commented on Quiz-v01/src/test/java/com/github/kondury/quiz/io/ConsoleOutputServiceTest.java:
```diff
@@ -0,0 +1,35 @@
+
+class ConsoleOutputServiceTest {
+
+    ConsoleOutputUtils utils = new ConsoleOutputUtils();
+
+    OutputService outputService;
```
`private`

kondury replied:
сделано

stvort commented on Quiz-v01/src/test/java/com/github/kondury/quiz/io/ConsoleOutputUtils.java:
```diff
@@ -0,0 +1,29 @@
+import java.io.PrintStream;
+
+public class ConsoleOutputUtils {
+    private final ByteArrayOutputStream outContent = new ByteArrayOutputStream();
+    private final ByteArrayOutputStream errContent = new ByteArrayOutputStream();
+    private final PrintStream originalOut = System.out;
+    private final PrintStream originalErr = System.err;
```
Вот если поток извне передавать, то подмена системного ввода-вывода в тестах будет не нужна

kondury replied:
планирую улучшить в hw 2

kondury replied:
сделано

kondury committed:
hw 1 made corrections according review
Answer is now Record, not interface
Quiestion class is corrected accordingly review comments
ChoiceAnswer and TextAnswer are removed, enum QuestionType is added
ConsoleOutputService:
PrintStream is injected now with ConsoleOutputStreamProvider (wraps PrintStream bean)
removed CsvQuestionsLoader and moved its methods into CsvQuizDao class
ExtensiveQuestionFormatter is corrected according domain objects changes
in OutputService removed unused versions of output()
in OutputService refactored output() method to take into String argument
class Question.Factory is moved on upper level and renamed into QuestionFactory

stvort commented on Quiz-v01/src/main/java/com/github/kondury/quiz/dao/CsvQuizDao.java:
```diff
@@ -0,0 +1,49 @@
+                questions.add(QuestionFactory.createFromCsv(line));
+            }
+        } catch (IOException e) {
+            throw new RuntimeException("IO error during file access: %s".formatted(csvDaoFilename), e);
```
Лучше прямо свое исключение сделать

kondury replied:
сделано

stvort commented on Quiz-v01/src/test/java/com/github/kondury/quiz/io/ConsoleOutputServiceTest.java:
```diff
@@ -0,0 +1,34 @@
+
+    @BeforeEach
+    public void setUp() {
+        utils.setUpStreams();
```
Это не нужно

kondury replied:
убрано

stvort commented on Quiz-v01/src/test/java/com/github/kondury/quiz/io/ConsoleOutputServiceTest.java:
```diff
@@ -0,0 +1,34 @@
+    @BeforeEach
+    public void setUp() {
+        utils.setUpStreams();
+        var outputStreamProvider = new ConsoleOutputStreamProvider();
```
```
class ConsoleOutputServiceTest {

    private final ConsoleOutputUtils utils = new ConsoleOutputUtils();

    private OutputService outputService;
    private OutputStream outputStream;
    private PrintStream printStream;

    @BeforeEach
    public void setUp() {
        utils.setUpStreams();
        outputStream = new ByteArrayOutputStream();
        printStream = new PrintStream(outputStream);
        var outputStreamProvider = mock(OutputStreamProvider.class);
        when(outputStreamProvider.getPrintStream()).thenReturn(printStream);
        outputService = new ConsoleOutputService(outputStreamProvider);
    }

    @AfterEach
    public void tearDown() {
        utils.restoreStreams();
    }

    @Test
    void output_SendTextToOutput_OutputIsTheSame() {
        String text = "Test string";
        outputService.output(text);
        assertEquals(text + System.lineSeparator(), outputStream.toString());
    }
}
```

kondury replied:
исправлено

stvort commented on Quiz-v01/src/main/java/com/github/kondury/quiz/io/ConsoleOutputService.java:
```diff
@@ -0,0 +1,17 @@
+
+import java.io.PrintStream;
+
+public class ConsoleOutputService implements OutputService {
```
Console этот сервис или нет, зависит от реализации `OutputStreamProvider`

kondury replied:
переименовал сервис

stvort commented on Quiz-v01/src/main/java/com/github/kondury/quiz/io/ConsoleOutputService.java:
```diff
@@ -0,0 +1,17 @@
+    private final PrintStream out;
+
+    public ConsoleOutputService(OutputStreamProvider outputStreamProvider) {
+        this.out = outputStreamProvider.getPrintStream();
```
Чтобы идея, заложенная в провайдер, сработала, `getPrintStream` нужно дергать в момент вызова `output`

kondury replied:
исправлено

kondury commented:
added hw1
